himbaechel/uarch/gowin/pack.cc

```diff
@@ -618,21 +625,142 @@ struct GowinPacker
             break;
         }
         ci.setParam(ctx->id("INMODE"), in_mode);
-
-        if (ci.type == id_IOLOGICI_EMPTY) {
-            return;
-        }
-
         // disconnect D input: it is wired internally
         nets_to_remove.push_back(ci.getPort(in_port)->name);
         in_iob->disconnectPort(id_O);
         ci.disconnectPort(in_port);
+        if (ci.type == id_IOLOGICI_EMPTY) {
+            ci.movePortTo(id_Q, in_iob, id_O);
+            return;
+        }
+
         set_daaj_nets(ci, iob_bel);
         reconnect_ides_outs(&ci);
 
         make_iob_nets(*in_iob);
     }
 
+    void pack_iodelay()
+    {
+        log_info("Pack IODELAY...\n");
+        std::vector<IdString> cells_to_remove;
+        std::vector<IdString> nets_to_remove;
+        std::vector<std::unique_ptr<CellInfo>> new_cells;
+
+        for (auto &cell : ctx->cells) {
+            CellInfo &ci = *cell.second;
+            if (ci.type != id_IODELAY) {
+                continue;
+            }
+            if (ctx->debug) {
+                log_info("pack %s of type %s.\n", ctx->nameOf(&ci), ci.type.c_str(ctx));
+            }
+            // There is only one delay line in the IO block, which can be either
+            // input or output.  Define which case we are dealing with.
+            bool is_idelay = false;
+            NetInfo *di_net = ci.ports.at(id_DI).net;
+            NetInfo *do_net = ci.ports.at(id_DO).net;
+            CellInfo *iob = net_driven_by(ctx, di_net, is_iob, id_O);
+            if (iob != nullptr) {
+                NPNR_ASSERT(iob->bel != BelId());
+                NPNR_ASSERT_MSG(di_net->users.entries() == 1, "IODELAY should be the only sink in the network.\n");
```

Comment:
This is better as a log_error, as it's something the user could conceivably trigger with a design